examples/user_guide/Performance_and_Debugging.ipynb

```diff
@@ -46,6 +46,80 @@
+    "\n",
+    "When to use which approach cannot be answered easily and is never completely clear cut. As a general guide however use `asyncio` when you are performing many slow IO bound operations concurrently (on the order of 100 and more) and use threading when you are performing a limited number (less than the number of available threads). In some situations threading can also be useful for CPU intensive operations where the operation being executed [releases the GIL](https://realpython.com/async-io-python/), this includes many NumPy, Pandas and Numba functions.\n",
+    "\n",
+    "### Threading\n",
+    "\n",
+    "Threading in Panel can be enabled easily by setting the `config.nthreads` parameter. This will start a `ThreadPoolExecutor` with the specified number of threads (or if set to `0` it will set the number of threads based on your system, i.e. `min(32, os.cpu_count() + 4)`). Whenever an event is generated or a periodic callback fires Panel will then automatically dispatch the event to the executor. If launching an application with `panel serve` you can configure this option with the `--num-threads` option.\n",
+    "\n",
+    "### AsyncIO\n",
+    "\n",
+    "When using Python>=3.8 you can use async callbacks wherever you would ordinarily use a regular synchronous function. For instance you can use `pn.bind` on an async function:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import aiohttp\n",
+    "\n",
+    "widget = pn.widgets.IntSlider(start=0, end=10)\n",
+    "\n",
+    "async def get_img(index):\n",
+    "    async with aiohttp.ClientSession() as session:\n",
+    "        async with session.get(f\"https://picsum.photos/800/300?image={index}\") as resp:\n",
+    "            return pn.pane.JPG(await resp.read())\n",
+    "            \n",
+    "pn.Column(widget, pn.bind(get_img, widget))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this example Panel will invoke the function and update the output when the function returns while leaving the process unblocked for the duration of the `aiohttp` request. \n",
+    "\n",
+    "Similarly you can attach asynchronous callbacks using `.param.watch`:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "widget = pn.widgets.IntSlider(start=0, end=10)\n",
+    "\n",
+    "image = pn.pane.JPG()\n",
+    "\n",
+    "async def update_img(event):\n",
+    "    async with aiohttp.ClientSession() as session:\n",
+    "        async with session.get(f\"https://picsum.photos/800/300?image={event.new}\") as resp:\n",
+    "            image.object = await resp.read()\n",
+    "            \n",
+    "widget.param.watch(update_img, 'value')\n",
+    "widget.param.trigger('value')\n",
+    "            \n",
+    "pn.Column(widget, image)"
```

Comment:
Also unable to run this snippet either in a notebook or in a script.

When I launch the app from a script I get the following error:
```
2021-12-01 18:04:04,067 Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x7fa69897b250>>, <Task finished name='Task-8' coro=<_needs_document_lock.<locals>._needs_document_lock_wrapper() done, defined at /Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/server/session.py:77> exception=TypeError("argument should be a bytes-like object or ASCII string, not 'NoneType'")>)
Traceback (most recent call last):
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/tornado/ioloop.py", line 765, in _discard_future_result
    future.result()
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/server/session.py", line 93, in _needs_document_lock_wrapper
    result = func(self, *args, **kwargs)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/server/session.py", line 227, in with_document_locked
    return func(*args, **kwargs)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/document/callbacks.py", line 450, in wrapper
    return invoke_with_curdoc(doc, invoke)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/document/callbacks.py", line 408, in invoke_with_curdoc
    return f()
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/document/callbacks.py", line 449, in invoke
    return f(*args, **kwargs)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/document/callbacks.py", line 179, in remove_then_invoke
    return callback()
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/panel/pane/base.py", line 153, in _update_object
    self._update(ref, old_model)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/panel/pane/markup.py", line 48, in _update
    model.update(**self._get_properties())
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/panel/pane/image.py", line 117, in _get_properties
    data = base64.b64decode(data)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/base64.py", line 80, in b64decode
    s = _bytes_from_decode_data(s)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/base64.py", line 45, in _bytes_from_decode_data
    raise TypeError("argument should be a bytes-like object or ASCII "
TypeError: argument should be a bytes-like object or ASCII string, not 'NoneType'
```

Reply:
Requires master, which version are you running?

Reply:
I run them from this branch, `enable_threading`.

Reply:
Guess it needs to be rebased.